### ISSUE TITLE
Acess violation being thrown when my class calls a function.

### ISSUE THREAD

**Issue by Snorflake** [vinniefalco/LuaBridge#118]
I have this function
```
	Vector WorldToScreen(const Vector origin) {
		Vector screen(0,0,0);
		/*SOMETHING IN HERE CRASHES*/
		if (!ScreenTransform(origin, screen)) {
                       //Do stuff here
		}
		/*END*/
		return screen;
	}
```

That call crashes. Why? Because of this.
```

	bool ScreenTransform(const Vector& point, Vector& screen) {
		 matrix3x4_t w2sMatrix = g_pEngine->WorldToScreenMatrix();

```

that's literally the only thing in there (and a return true). The error is in lua_pcall (static libs, so im not 100% sure where). 
My last error message: `Exception thrown at 0x33D9C121 (the dll.dll) in csgo.exe: 0xC0000005: Access violation reading location 0x8D050014.
`

**Comment by josh33901** [comment]
if that is your c++ code, then it shouldn't have anything to do with luabridge.

**Comment by Snorflake** [comment]
It works fine when calling through C++

**Comment by josh33901** [comment]
what is the lua code?

**Comment by Snorflake** [comment]
Game.Utils:WorldToScreen(pos)

**Comment by Dan782** [comment]
wrong index you dip

**Comment by Snorflake** [comment]
actually it isnt

**Comment by Snorflake** [comment]
callstack: http://www.snorflake.com/u/Vnmh.png
Exception thrown at 0x339DC1A1 (CS-Lua.dll) in csgo.exe: 0xC0000005: Access violation reading location 0x00000014.

**Comment by m1k0s** [comment]
Try changing:

Vector WorldToScreen(const Vector origin)

to either:

Vector WorldToScreen(const Vector& origin)

or:

Vector WorldToScreen(Vector origin)


On 15 December 2015 at 05:59, Snorflake <notifications@github.com> wrote:

> callstack: http://www.snorflake.com/u/Vnmh.png
> Exception thrown at 0x339DC1A1 (CS-Lua.dll) in csgo.exe: 0xC0000005:
> Access violation reading location 0x00000014.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vinniefalco/LuaBridge/issues/118#issuecomment-164651823>
> .
>



-- 

*steliaros*.com

**Comment by Snorflake** [comment]
the function is currently
```
	Vector WorldToScreen(lua_State* L) {
		Vector screen(0,0,0);
		Vector origin = luabridge::LuaRef::fromStack(L, 2).cast<Vector>();  //Vector is correct

		matrix3x4_t w2sMatrix = g_pEngine->WorldToScreenMatrix();
		/*SOMETHING IN HERE CRASHES*/
		//if (!ScreenTransform(origin, screen)) {
			/*int iScreenWidth, iScreenHeight;
			g_pEngine->GetScreenSize(iScreenWidth, iScreenHeight);

			float fScreenX = iScreenWidth / 2.0f;
			float fScreenY = iScreenHeight / 2.0f;

			fScreenX += 0.5f * screen.x * iScreenWidth + 0.5f;
			fScreenY -= 0.5f * screen.y * iScreenHeight + 0.5f;

			screen.x = fScreenX;
			screen.y = fScreenY;*/
		//}
		/*END*/
		return screen;
	}
```

The function only crashes when i call `matrix3x4_t w2sMatrix = g_pEngine->WorldToScreenMatrix();`

**Comment by m1k0s** [comment]
how are you registering this with luabridge?

On 15 December 2015 at 11:59, Snorflake <notifications@github.com> wrote:

> the function is currently
>
>     Vector WorldToScreen(lua_State* L) {
>         Vector screen(0,0,0);
>         Vector origin = luabridge::LuaRef::fromStack(L, 2).cast<Vector>();  //Vector is correct
>
>         matrix3x4_t w2sMatrix = g_pEngine->WorldToScreenMatrix();
>         /*SOMETHING IN HERE CRASHES*/
>         //if (!ScreenTransform(origin, screen)) {
>             /*int iScreenWidth, iScreenHeight;
>             g_pEngine->GetScreenSize(iScreenWidth, iScreenHeight);
>
>             float fScreenX = iScreenWidth / 2.0f;
>             float fScreenY = iScreenHeight / 2.0f;
>
>             fScreenX += 0.5f * screen.x * iScreenWidth + 0.5f;
>             fScreenY -= 0.5f * screen.y * iScreenHeight + 0.5f;
>
>             screen.x = fScreenX;
>             screen.y = fScreenY;*/
>         //}
>         /*END*/
>         return screen;
>     }
>
> The function only crashes when i call matrix3x4_t w2sMatrix =
> g_pEngine->WorldToScreenMatrix();
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vinniefalco/LuaBridge/issues/118#issuecomment-164742171>
> .
>



-- 

*steliaros*.com

**Comment by Snorflake** [comment]
```

			.beginClass<LUAUtils>("__Utils")
				.addFunction("WorldToScreen", &LUAUtils::WorldToScreen)
				.addFunction("IsPlayer", &LUAUtils::IsPlayer)
				.addFunction("GetHitboxPos", &LUAUtils::GetHitboxPosition)
			.endClass()

```

and then i have a global variable LUAUtils g_Utils; and register this...
`
			.addVariable("Utils", &g_Utils, false)`

**Comment by josh33901** [comment]
if you are using it with the lua state, isnt it a CFunc rather than a regular one?

**Comment by m1k0s** [comment]
All functions can have an optional lua_State* as the last param, so this
*should* work, although not how I'd do it (also a CFunc would need to
manually push the return onto the lua stack and return 1)... L is not used
in the body so the sig could easily be:

Vector WorldToScreen(const Vector& origin);

Hard to say why you are crashing - on the face of it it does not
appear to be luabridge related...


On 15 December 2015 at 18:48, josh33901 <notifications@github.com> wrote:

> if you are using it with the lua state, isnt it a CFunc rather than a
> regular one?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vinniefalco/LuaBridge/issues/118#issuecomment-164852973>
> .
>



-- 

*steliaros*.com

**Comment by Snorflake** [comment]
But it has to be (I was using L, its not that that's crashing). Calling
that function from the same place/thread etc in c++ works as intended...
I've even mutex'd the lua_state to make sure its not being weird with
threads.

**Comment by m1k0s** [comment]
Wait. Threads? To ask the obvious, you are guaranteeing that anything to do
with the Lua side is done on the one thread, right? (Or that you have a
separate lua_State per thread?)

On Wednesday, 16 December 2015, Snorflake <notifications@github.com> wrote:

> But it has to be (I was using L, its not that that's crashing). Calling
> that function from the same place/thread etc in c++ works as intended...
> I've even mutex'd the lua_state to make sure its not being weird with
> threads.
>
> On Wed, 16 Dec 2015 3:39 am m1k0s <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
> > All functions can have an optional lua_State* as the last param, so this
> > *should* work, although not how I'd do it (also a CFunc would need to
> > manually push the return onto the lua stack and return 1)... L is not
> used
> > in the body so the sig could easily be:
> >
> > Vector WorldToScreen(const Vector& origin);
> >
> > Hard to say why you are crashing - on the face of it it does not
> > appear to be luabridge related...
> >
> >
> > On 15 December 2015 at 18:48, josh33901 <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
> >
> > > if you are using it with the lua state, isnt it a CFunc rather than a
> > > regular one?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> >
> https://github.com/vinniefalco/LuaBridge/issues/118#issuecomment-164852973
> > >
> > > .
> > >
> >
> >
> >
> > --
> >
> > *steliaros*.com
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/vinniefalco/LuaBridge/issues/118#issuecomment-164869897
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vinniefalco/LuaBridge/issues/118#issuecomment-164941099>
> .
>


-- 

*steliaros*.com

**Comment by Snorflake** [comment]
Yes I am. I've got it to stop crashing by passing the matrix3x4_t as an
argument and binding a GetMatrix() function. the matrix3x4_t is NaN but I
think that's my problem.

**Comment by Snorflake** [comment]
Alright - update. I just made a ToScreen function in the vector class. I have NO clue why I was getting an access violation in my utils class but whatever, it's fine now.
